server/src/main/java/org/elasticsearch/index/reindex/ReindexJobState.java

```diff
@@ -0,0 +1,128 @@
+    private static String REINDEX_RESPONSE = "reindex_response";
+    private static String REINDEX_EXCEPTION = "reindex_exception";
+
+    static {
+        PARSER.declareString(ConstructingObjectParser.constructorArg(),  new ParseField(EPHEMERAL_TASK_ID));
+        PARSER.declareObject(ConstructingObjectParser.optionalConstructorArg(), (p, c) -> BulkByScrollResponse.fromXContent(p),
+            new ParseField(REINDEX_RESPONSE));
+        PARSER.declareObject(ConstructingObjectParser.optionalConstructorArg(), (p, c) -> ElasticsearchException.fromXContent(p),
+            new ParseField(REINDEX_EXCEPTION));
+    }
+
+    private final TaskId taskId;
+    private final BulkByScrollResponse reindexResponse;
+    private final ElasticsearchException jobException;
+
+    private ReindexJobState(String taskId, BulkByScrollResponse reindexResponse, ElasticsearchException jobException) {
+        this(new TaskId(taskId), reindexResponse, jobException);
+    }
+
+    ReindexJobState(TaskId taskId, @Nullable BulkByScrollResponse reindexResponse, @Nullable ElasticsearchException jobException) {
+        this.taskId = taskId;
+        assert (reindexResponse == null) || (jobException == null) : "Either response or exception must be null";
+        this.reindexResponse = reindexResponse;
+        this.jobException = jobException;
+    }
+
+    public ReindexJobState(StreamInput in) throws IOException {
+        taskId = TaskId.readFromStream(in);
+        reindexResponse = in.readOptionalWriteable((input) -> {
+            BulkByScrollResponse response = new BulkByScrollResponse();
+            response.readFrom(input);
+            return response;
+        });
+        jobException = in.readException();
+    }
+
+    @Override
+    public String getWriteableName() {
+        return NAME;
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        taskId.writeTo(out);
+        out.writeOptionalWriteable(reindexResponse);
+        out.writeException(jobException);
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
+        builder.field(EPHEMERAL_TASK_ID, taskId.toString());
+        if (reindexResponse != null) {
+            builder.field(REINDEX_RESPONSE);
+            builder.startObject();
+            reindexResponse.toXContent(builder, params);
+            builder.endObject();
+        }
+        if (jobException != null) {
+            builder.field(REINDEX_EXCEPTION);
+            builder.startObject();
+            jobException.toXContent(builder, params);
+            builder.endObject();
+        }
+        return builder.endObject();
+    }
+
+    public BulkByScrollResponse getReindexResponse() {
+        return reindexResponse;
+    }
+
+    public ElasticsearchException getJobException() {
+        return jobException;
+    }
+
+    public TaskId getTaskId() {
```

Comment:
Maybe rename to getEphemeralTaskId(), it could make call-site code a little easier to read (at least I mixed it up with the persistent task id).